File: day2.cpp

#include <fstream>
#include <sstream>
#include <string>

#include <vector>
#include <set>

std::vector<std::string> valuesFromFile(const char * path)
{
    std::vector<std::string> values;

    std::ifstream input;
    input.open(path);

    std::string line;
    while (std::getline(input, line))
    {
        values.push_back(line);
    }
    input.close();

    return values;
}

int main()
{
    auto values = valuesFromFile("day2.txt");

    {
        int twoCount = 0;
        int threeCount = 0;

        for (auto && value : values)
        {
            int counts[26] = { 0 };
            for (char ch : value)
            {
                counts[ch - 'a']++;
            }

            bool hasTwoCount = false;
            bool hasThreeCount = false;
            for (int count : counts)
            {
                if (count == 2)
                {
                    hasTwoCount = true;
                }
                else if (count == 3)
                {
                    hasThreeCount = true;
                }
            }

            if (hasTwoCount)
            {
                twoCount++;
            }
            if (hasThreeCount)
            {
                threeCount++;
            }
        }

        printf("Score: %d\n", twoCount * threeCount);
    }

    std::getc(stdin);
}

File: day2.js
var fs = require('fs');

var values = fs.readFileSync('day2.txt', 'utf8').split('\n');

function calculateChecksum(values)
{
    var twoCount = 0;
    var threeCount = 0;

    values.forEach(value => {
        var counts = new Object();

        for (var i = 0; i < value.length; i++) {
            if (counts[value.charAt(i)] == null)
            {
                counts[value.charAt(i)] = 0;
            }
            counts[value.charAt(i)]++;
        }

        var hasTwoCount = false;
        var hasThreeCount = false;
        Object.keys(counts).forEach(char => {
            if (counts[char] == 2)
            {
                hasTwoCount = true;
            }
            else if (counts[char] == 3)
            {
                hasThreeCount = true;
            }
        });

        if (hasTwoCount)
        {
            twoCount++;
        }
        if (hasThreeCount)
        {
            threeCount++;
        }
    });

    return twoCount * threeCount;
}

console.log(calculateChecksum(values));